app/src/chat_engine.py

```diff
@@ -0,0 +1,146 @@
+import logging
+from abc import ABC, abstractmethod
+from typing import Type
+
+import chainlit as cl
+import src.adapters.db as db
+from chainlit.input_widget import Slider, Switch
+from src.db.models.document import Chunk
+from src.format import format_guru_cards
+from src.generate import generate
+from src.retrieve import retrieve
+from src.shared import get_embedding_model
+
+logger = logging.getLogger(__name__)
+
+
+class ChatEngineInterface(ABC):
+    engine_id: str
+    name: str
+
+    @abstractmethod
+    async def on_start(self) -> dict:
+        pass
+
+    @abstractmethod
+    def on_message(self, question: str, cl_message: cl.Message) -> dict:
+        pass
+
+    @abstractmethod
+    def format_answer_message(self, results: dict) -> str:
+        pass
+
+
+def all_subclasses(cls: Type) -> set:
+    return {cls}.union(s for c in cls.__subclasses__() for s in all_subclasses(c))
+
+
+def available_engines() -> list[str]:
+    return [
+        engine_class.engine_id
+        for engine_class in all_subclasses(ChatEngineInterface)
+        if hasattr(engine_class, "engine_id") and engine_class.engine_id
+    ]
+
+
+def create_engine(engine_id: str) -> ChatEngineInterface | None:
+    if engine_id not in available_engines():
+        return None
+
+    chat_engine_class = next(
+        engine_class
+        for engine_class in all_subclasses(ChatEngineInterface)
+        if hasattr(engine_class, "engine_id") and engine_class.engine_id == engine_id
+    )
+    return chat_engine_class()
+
+
+# Subclasses of ChatEngineInterface can be extracted into a separate file if it gets too large
```

Comment:
👍  yeah I imagine in the future we'd want to put these under something like /app/src/engines